src/vscode/superbol-vscode-platform/superbol_cfg_explorer.ml

```diff
@@ -0,0 +1,284 @@
+  s
+
+let _log message = ignore(Window.showInformationMessage () ~message)
+
+(* DECORATION TYPE *)
+
+let decorationType =
+  let backgroundColor = Ojs.string_to_js "#75ff3388" in
+  let options = Ojs.obj [|("backgroundColor", backgroundColor)|] in
+  Window.createTextEditorDecorationType ~options
+
+(* GRAPH TYPE*)
+
+let typ_of_string = function
+  | "d3" -> `D3
+  | _ -> `Dot
+
+(* GRAPH FROM LSP *)
+
+type graph = {
+  string_repr_dot: string;
+  string_repr_d3: string;
+  nodes_pos: (string * Jsonoo.t) list;
+  name: string;
+}
+let decode_graph res =
+  let string_repr_dot =
+    Jsonoo.Decode.field "string_repr_dot" Jsonoo.Decode.string res in
+  let string_repr_d3 =
+    Jsonoo.Decode.field "string_repr_d3" Jsonoo.Decode.string res in
+  let nodes_pos = Jsonoo.Decode.field "nodes_pos" Jsonoo.Decode.(dict id) res in
+  let nodes_pos = Hashtbl.to_seq nodes_pos |> List.of_seq in
+  let name = Jsonoo.Decode.field "name" Jsonoo.Decode.string res in
+  { name; nodes_pos; string_repr_dot; string_repr_d3 }
+
+
+(* WEBVIEW MANAGEMENT *)
+
+let webview_panels = Hashtbl.create 1
+let window_listener = ref None
+
+let webviewpanel_disposal ~filename () =
+  Hashtbl.remove webview_panels filename;
+  if Hashtbl.length webview_panels == 0
+  then (
+    Option.iter Disposable.dispose !window_listener;
+    window_listener := None);
+  match Window.activeTextEditor () with
+  | None -> ()
+  | Some text_editor ->
+    let uri = TextEditor.document text_editor
+              |> TextDocument.uri in
+    if String.equal filename @@ Uri.path uri
+    then TextEditor.setDecorations text_editor
+        ~decorationType ~rangesOrOptions:(`Ranges [])
+
+let create_or_get_webview ~graph ~uri =
+  let filename = Uri.path uri in
+  match Hashtbl.find_opt webview_panels filename with
+  | Some (webview_panel, _) ->
+    WebviewPanel.reveal webview_panel ();
+    Hashtbl.replace webview_panels filename (webview_panel, graph);
+    WebviewPanel.webview webview_panel, false
+  | None ->
+    let webview_panel = Window.createWebviewPanel
+        ~viewType:"CFG" ~title:"COBOL CFG Viewer"
+        ~showOptions:(ViewColumn.Beside) in
+    let _ : Disposable.t =
+      WebviewPanel.onDidDispose webview_panel ()
+        ~listener:(webviewpanel_disposal ~filename)
+        ~thisArgs:Ojs.null ~disposables:[] in
+    let webview = WebviewPanel.webview webview_panel in
+    WebView.set_options webview (WebviewOptions.create ~enableScripts:true ());
+    Hashtbl.add webview_panels filename (webview_panel, graph);
+    webview, true
+
+let webview_n_graph_find_opt ~uri =
+  Hashtbl.find_opt webview_panels @@ Uri.path uri
+  |> Option.map begin fun (w,g) -> WebviewPanel.webview w, g end
+
+let update_graph ~uri graph =
+  let filename = Uri.path uri in
+  match Hashtbl.find_opt webview_panels filename with
+  | Some (wvp, _) ->
+    Hashtbl.replace webview_panels filename (wvp, graph)
+  | None -> ()
+
+(* CLICK ON NODE *)
+
+let ignore_next_selection_change = ref false
```

Comment:
What is this state attached to? a window? an editor? a webview?

Reply:
This has been reworked